File: packages/python/tests/test_predict.py
"""
Tests for taxonbodymassml.predict_mass().

Tests that require model artifacts (the ~2 GB download) are skipped
unless TAXONBODYMASSML_RUN_INTEGRATION=1 is set in the environment.
"""

import os
import warnings

import pandas as pd
import pytest

INTEGRATION = os.environ.get("TAXONBODYMASSML_RUN_INTEGRATION", "0") == "1"
skip_without_artifacts = pytest.mark.skipif(
    not INTEGRATION,
    reason="Requires model artifacts; set TAXONBODYMASSML_RUN_INTEGRATION=1",
)


# ---------------------------------------------------------------------------
# Unit tests (no artifacts required)
# ---------------------------------------------------------------------------
def test_resolve_ci_level_false():
    from taxonbodymassml._predict import _resolve_ci_level

    assert _resolve_ci_level(False) is None


def test_resolve_ci_level_true():
    from taxonbodymassml._predict import _resolve_ci_level

    assert _resolve_ci_level(True) == 0.90


def test_resolve_ci_level_float():
    from taxonbodymassml._predict import _resolve_ci_level

    assert _resolve_ci_level(0.80) == pytest.approx(0.80)


def test_resolve_ci_level_invalid():
    from taxonbodymassml._predict import _resolve_ci_level

    with pytest.raises(ValueError):
        _resolve_ci_level(1.5)
    with pytest.raises(ValueError):
        _resolve_ci_level(0.0)


def test_predict_unknown_method():
    import taxonbodymassml as tbm

    with pytest.raises(ValueError, match="Unknown method"):
        tbm.predict_mass("Mus musculus", method="RandomForest")


def test_predict_dataframe_missing_columns():
    import taxonbodymassml as tbm

    with pytest.raises(ValueError, match="missing taxonomy columns"):
        tbm.predict_mass(pd.DataFrame({"kingdom": ["Animalia"]}))


# ---------------------------------------------------------------------------
# Integration tests (require artifacts + network)
# ---------------------------------------------------------------------------
@skip_without_artifacts
def test_predict_single_species():
    import taxonbodymassml as tbm

    result = tbm.predict_mass("Mus musculus")
    assert isinstance(result, pd.DataFrame)
    assert "mass_g" in result.columns
    assert result["mass_g"].iloc[0] > 0


@skip_without_artifacts
def test_predict_confidence_interval_true():
    import taxonbodymassml as tbm

    result = tbm.predict_mass("Mus musculus", confidence_interval=True)
    assert "lower_bound" in result.columns
    assert "upper_bound" in result.columns
    assert result["confidence"].iloc[0] == pytest.approx(0.90)
    assert result["lower_bound"].iloc[0] < result["mass_g"].iloc[0]
    assert result["upper_bound"].iloc[0] > result["mass_g"].iloc[0]


@skip_without_artifacts
def test_predict_confidence_interval_custom():
    import taxonbodymassml as tbm

    r90 = tbm.predict_mass("Mus musculus", confidence_interval=0.90)
    r80 = tbm.predict_mass("Mus musculus", confidence_interval=0.80)
    width90 = r90["upper_bound"].iloc[0] - r90["lower_bound"].iloc[0]
    width80 = r80["upper_bound"].iloc[0] - r80["lower_bound"].iloc[0]
    assert width90 > width80  # wider interval at higher coverage


@skip_without_artifacts
def test_predict_include_taxonomy():
    import taxonbodymassml as tbm

    result = tbm.predict_mass("Mus musculus", include_taxonomy=True)
    assert "kingdom" in result.columns
    assert "species_resolved" in result.columns


@skip_without_artifacts
def test_predict_unresolvable_species():
    import taxonbodymassml as tbm

    with warnings.catch_warnings(record=True) as w:
        warnings.simplefilter("always")
        result = tbm.predict_mass("xxxxnotaspeciesxxxx")
    assert any("No species could be resolved" in str(warning.message) for warning in w)
    import math

    assert math.isnan(result["mass_g"].iloc[0])


@skip_without_artifacts
def test_predict_list():
    import taxonbodymassml as tbm

    result = tbm.predict_mass(["Mus musculus", "Panthera leo"])
    assert len(result) == 2
    assert all(result["mass_g"] > 0)


@skip_without_artifacts
def test_predict_taxonomy_dataframe_input():
    import taxonbodymassml as tbm

    tax = tbm.lookup_taxonomy("Mus musculus")
    result = tbm.predict_mass(tax)
    assert result["mass_g"].iloc[0] > 0


# ---------------------------------------------------------------------------
# fuzzy_match_name column semantics
# ---------------------------------------------------------------------------


@skip_without_artifacts
def test_fuzzy_match_corrected_name_in_species_original_in_matched_name():
    import taxonbodymassml as tbm

    result = tbm.predict_mass("Ballanus glandula", fuzzy_match_name=True)
    assert result["taxon"].iloc[0] == "Balanus glandula"
    assert result["matched_name"].iloc[0] == "Ballanus glandula"
    assert "matched_name" in result.columns


@skip_without_artifacts
def test_fuzzy_match_matched_name_none_when_no_correction_needed():
    import taxonbodymassml as tbm

    result = tbm.predict_mass("Balanus glandula", fuzzy_match_name=True)
    assert result["taxon"].iloc[0] == "Balanus glandula"
    assert result["matched_name"].iloc[0] is None
    assert "matched_name" in result.columns


@skip_without_artifacts
def test_fuzzy_match_species_none_and_matched_name_set_when_gbif_finds_no_match():
    import math

    import taxonbodymassml as tbm

    with warnings.catch_warnings(record=True):
        warnings.simplefilter("always")
        result = tbm.predict_mass(
            "Xyzzy_definitely_not_a_species_12345", fuzzy_match_name=True
        )  # noqa: E501
    assert result["taxon"].iloc[0] is None
    assert result["matched_name"].iloc[0] == "Xyzzy_definitely_not_a_species_12345"
    assert math.isnan(result["mass_g"].iloc[0])


# ---------------------------------------------------------------------------
# Dictionary lookup and include_source
# ---------------------------------------------------------------------------


@skip_without_artifacts
def test_dict_hit_returns_empirical_mass():
    """Nucella ostrina is in training data with mass_g=0.7."""
    import taxonbodymassml as tbm

    result = tbm.predict_mass("Nucella ostrina")
    assert result["mass_g"].iloc[0] == pytest.approx(0.7)


@skip_without_artifacts
def test_include_source_dict_hit_returns_source_string():
    import taxonbodymassml as tbm

    result = tbm.predict_mass("Nucella ostrina", include_source=True)
    assert "source" in result.columns
    assert result["source"].iloc[0] == "Novak_unpubl"


@skip_without_artifacts
def test_dict_hit_ci_columns_are_nan():
    import math

    import taxonbodymassml as tbm

    result = tbm.predict_mass("Nucella ostrina", confidence_interval=True)
    assert "lower_bound" in result.columns
    assert math.isnan(result["lower_bound"].iloc[0])
    assert math.isnan(result["upper_bound"].iloc[0])
    assert math.isnan(result["confidence"].iloc[0])


@skip_without_artifacts
def test_include_source_model_inferred_genus_known():
    """Nucella lima is not in training data but the genus Nucella is known."""
    import taxonbodymassml as tbm

    result = tbm.predict_mass("Nucella lima", include_source=True)
    assert "source" in result.columns
    assert result["source"].iloc[0] == "tbmML_genus"


@skip_without_artifacts
def test_include_source_unresolvable_returns_none():
    import taxonbodymassml as tbm

    with warnings.catch_warnings(record=True):
        warnings.simplefilter("always")
        result = tbm.predict_mass(
            "Xyzzy_definitely_not_a_species_12345", include_source=True
        )  # noqa: E501
    assert "source" in result.columns
    assert result["source"].iloc[0] is None


@skip_without_artifacts
def test_mixed_dict_model_unresolved_order_preserved():
    """Output order is preserved across dict hits, model rows, and unresolved."""
    import math

    import taxonbodymassml as tbm

    sp = [
        "Nucella ostrina",  # dict hit
        "Canis lupus",  # model-inferred
        "Xyzzy_definitely_not_a_species_12345",  # unresolvable
    ]
    with warnings.catch_warnings(record=True):
        warnings.simplefilter("always")
        result = tbm.predict_mass(sp, include_source=True)

    assert len(result) == 3
    assert result["taxon"].iloc[0] == "Nucella ostrina"
    assert result["mass_g"].iloc[0] == pytest.approx(0.7)
    assert result["source"].iloc[0] == "Novak_unpubl"
    assert result["mass_g"].iloc[1] > 0
    assert result["source"].iloc[1].startswith("tbmML_")
    assert math.isnan(result["mass_g"].iloc[2])
    assert result["source"].iloc[2] is None


# ---------------------------------------------------------------------------
# lookup parameter
# ---------------------------------------------------------------------------


@skip_without_artifacts
def test_lookup_false_routes_dict_species_through_model():
    """With lookup=False, a dict species should get model mass, not empirical."""
    import taxonbodymassml as tbm

    result = tbm.predict_mass("Nucella ostrina", lookup=False)
    assert result["mass_g"].iloc[0] != pytest.approx(0.7)


@skip_without_artifacts
def test_lookup_false_include_source_returns_tbmml_prefix():
    import taxonbodymassml as tbm

    result = tbm.predict_mass("Nucella ostrina", lookup=False, include_source=True)
    assert result["source"].iloc[0].startswith("tbmML_")

File: packages/r/tests/testthat/test-predict.R
test_that(".resolve_ci_level returns NULL for FALSE", {
  expect_null(TaxonBodyMassML:::.resolve_ci_level(FALSE))
})

test_that(".resolve_ci_level returns 0.90 for TRUE", {
  expect_equal(TaxonBodyMassML:::.resolve_ci_level(TRUE), 0.90)
})

test_that(".resolve_ci_level returns the numeric value for a float in (0, 1)", {
  expect_equal(TaxonBodyMassML:::.resolve_ci_level(0.80), 0.80)
  expect_equal(TaxonBodyMassML:::.resolve_ci_level(0.50), 0.50)
})

test_that(".resolve_ci_level errors on invalid values", {
  expect_error(TaxonBodyMassML:::.resolve_ci_level(0),   "confidence_interval")
  expect_error(TaxonBodyMassML:::.resolve_ci_level(1),   "confidence_interval")
  expect_error(TaxonBodyMassML:::.resolve_ci_level(1.5), "confidence_interval")
  expect_error(TaxonBodyMassML:::.resolve_ci_level(-0.1),"confidence_interval")
  expect_error(TaxonBodyMassML:::.resolve_ci_level("bad"),"confidence_interval")
})

test_that(".resolve_interval_method returns valid methods unchanged", {
  expect_equal(TaxonBodyMassML:::.resolve_interval_method("stratified"), "stratified")
  expect_equal(TaxonBodyMassML:::.resolve_interval_method("pooled"),     "pooled")
})

test_that(".resolve_interval_method errors on invalid values", {
  expect_error(TaxonBodyMassML:::.resolve_interval_method("marginal"), "interval_method")
  expect_error(TaxonBodyMassML:::.resolve_interval_method(""),         "interval_method")
  expect_error(TaxonBodyMassML:::.resolve_interval_method(TRUE),       "interval_method")
})

test_that("predict_mass() errors on unknown method", {
  expect_error(
    TaxonBodyMassML::predict_mass("Homo sapiens", method = "NotAModel"),
    "Unknown method"
  )
})

test_that("predict_mass() errors on invalid interval_method", {
  expect_error(
    TaxonBodyMassML::predict_mass("Homo sapiens", interval_method = "marginal"),
    "interval_method"
  )
})

test_that("predict_mass() errors when data.frame is missing required columns", {
  bad_df <- data.frame(kingdom = "Animalia", stringsAsFactors = FALSE)
  expect_error(
    TaxonBodyMassML::predict_mass(bad_df),
    "missing columns"
  )
})

test_that("predict_mass() errors when data.frame has some but not all required columns", {
  partial_df <- data.frame(
    kingdom  = "Animalia",
    phylum   = "Chordata",
    class    = "Mammalia",
    stringsAsFactors = FALSE
  )
  expect_error(
    TaxonBodyMassML::predict_mass(partial_df),
    "missing columns"
  )
})

# ---------------------------------------------------------------------------
# Integration tests — require downloaded artifacts
# ---------------------------------------------------------------------------

test_that("predict_mass() returns mass_g for a single species", {
  testthat::skip_if(!TaxonBodyMassML:::.artifacts_cached(),
                    "Model artifacts not cached; skipping integration test.")
  testthat::skip_if_offline()

  result <- TaxonBodyMassML::predict_mass("Homo sapiens")
  expect_s3_class(result, "data.frame")
  expect_true("mass_g" %in% names(result))
  expect_equal(nrow(result), 1L)
  expect_true(is.numeric(result$mass_g))
  expect_gt(result$mass_g, 0)
})

test_that("predict_mass() returns CI columns when confidence_interval = TRUE", {
  testthat::skip_if(!TaxonBodyMassML:::.artifacts_cached(),
                    "Model artifacts not cached; skipping integration test.")
  testthat::skip_if_offline()

  result <- TaxonBodyMassML::predict_mass("Canis lupus", confidence_interval = TRUE)
  expect_true(all(c("lower_bound", "upper_bound", "confidence") %in% names(result)))
  expect_equal(result$confidence, 0.90)
  expect_lt(result$lower_bound, result$mass_g)
  expect_gt(result$upper_bound, result$mass_g)
})

test_that("predict_mass() returns custom CI level", {
  testthat::skip_if(!TaxonBodyMassML:::.artifacts_cached(),
                    "Model artifacts not cached; skipping integration test.")
  testthat::skip_if_offline()

  result <- TaxonBodyMassML::predict_mass("Mus musculus", confidence_interval = 0.50)
  expect_equal(result$confidence, 0.50)
})

test_that("predict_mass() with include_taxonomy = TRUE includes taxonomy columns", {
  testthat::skip_if(!TaxonBodyMassML:::.artifacts_cached(),
                    "Model artifacts not cached; skipping integration test.")
  testthat::skip_if_offline()

  result <- TaxonBodyMassML::predict_mass("Panthera leo", include_taxonomy = TRUE)
  tax_cols <- c("kingdom", "phylum", "class", "order", "family", "genus",
                "species_resolved")
  expect_true(all(tax_cols %in% names(result)))
})

test_that("predict_mass() preserves input order for multiple species", {
  testthat::skip_if(!TaxonBodyMassML:::.artifacts_cached(),
                    "Model artifacts not cached; skipping integration test.")
  testthat::skip_if_offline()

  sp <- c("Homo sapiens", "Mus musculus", "Panthera leo")
  result <- TaxonBodyMassML::predict_mass(sp)
  expect_equal(result$taxon, sp)
})

test_that("predict_mass() accepts pre-resolved data.frame input", {
  testthat::skip_if(!TaxonBodyMassML:::.artifacts_cached(),
                    "Model artifacts not cached; skipping integration test.")
  testthat::skip_if_offline()

  tax <- TaxonBodyMassML::lookup_taxonomy("Canis lupus")
  result <- TaxonBodyMassML::predict_mass(tax)
  expect_s3_class(result, "data.frame")
  expect_true("mass_g" %in% names(result))
})

test_that("predict_mass() returns wider interval at higher coverage level", {
  testthat::skip_if(!TaxonBodyMassML:::.artifacts_cached(),
                    "Model artifacts not cached; skipping integration test.")
  testthat::skip_if_offline()

  r90 <- TaxonBodyMassML::predict_mass("Mus musculus", confidence_interval = 0.90)
  r80 <- TaxonBodyMassML::predict_mass("Mus musculus", confidence_interval = 0.80)
  width90 <- r90$upper_bound - r90$lower_bound
  width80 <- r80$upper_bound - r80$lower_bound
  expect_gt(width90, width80)
})

test_that("predict_mass() returns NA mass_g and warns for unresolvable species", {
  testthat::skip_if(!TaxonBodyMassML:::.artifacts_cached(),
                    "Model artifacts not cached; skipping integration test.")
  testthat::skip_if_offline()

  expect_warning(
    result <- TaxonBodyMassML::predict_mass("Xyzzy_definitely_not_a_species_12345"),
    "No species could be resolved"
  )
  expect_equal(nrow(result), 1L)
  expect_true(is.na(result$mass_g))
})

# ---------------------------------------------------------------------------
# fuzzy_match_name column semantics
# ---------------------------------------------------------------------------

test_that("predict_mass() with fuzzy_match_name: corrected name in species, original in matched_name", {
  testthat::skip_if(!TaxonBodyMassML:::.artifacts_cached(),
                    "Model artifacts not cached; skipping integration test.")
  testthat::skip_if_offline()

  result <- TaxonBodyMassML::predict_mass("Ballanus glandula",
                                          fuzzy_match_name = TRUE)
  expect_equal(result$taxon, "Balanus glandula")
  expect_equal(result$matched_name, "Ballanus glandula")
  expect_true("matched_name" %in% names(result))
})

test_that("predict_mass() with fuzzy_match_name: matched_name is NA when no correction needed", {
  testthat::skip_if(!TaxonBodyMassML:::.artifacts_cached(),
                    "Model artifacts not cached; skipping integration test.")
  testthat::skip_if_offline()

  result <- TaxonBodyMassML::predict_mass("Balanus glandula",
                                          fuzzy_match_name = TRUE)
  expect_equal(result$taxon, "Balanus glandula")
  expect_true(is.na(result$matched_name))
  expect_true("matched_name" %in% names(result))
})

test_that("predict_mass() with fuzzy_match_name: species NA and matched_name set when GBIF finds no match", {
  testthat::skip_if(!TaxonBodyMassML:::.artifacts_cached(),
                    "Model artifacts not cached; skipping integration test.")
  testthat::skip_if_offline()

  suppressWarnings(
    result <- TaxonBodyMassML::predict_mass("Xyzzy_definitely_not_a_species_12345",
                                            fuzzy_match_name = TRUE)
  )
  expect_true(is.na(result$taxon))
  expect_equal(result$matched_name, "Xyzzy_definitely_not_a_species_12345")
  expect_true(is.na(result$mass_g))
})

# ---------------------------------------------------------------------------
# Dictionary lookup and include_source
# ---------------------------------------------------------------------------

test_that("predict_mass() returns empirical mass for species in training data", {
  testthat::skip_if(!TaxonBodyMassML:::.artifacts_cached(),
                    "Model artifacts not cached; skipping integration test.")
  testthat::skip_if_offline()

  # Nucella ostrina is in training data with mass_g = 0.7, source = "Novak_unpubl"
  result <- TaxonBodyMassML::predict_mass("Nucella ostrina")
  expect_equal(result$mass_g, 0.7)
})

test_that("predict_mass() with include_source returns source for dictionary hit", {
  testthat::skip_if(!TaxonBodyMassML:::.artifacts_cached(),
                    "Model artifacts not cached; skipping integration test.")
  testthat::skip_if_offline()

  result <- TaxonBodyMassML::predict_mass("Nucella ostrina", include_source = TRUE)
  expect_true("source" %in% names(result))
  expect_equal(result$source, "Novak_unpubl")
})

test_that("predict_mass() dict hit with CI has NA bounds", {
  testthat::skip_if(!TaxonBodyMassML:::.artifacts_cached(),
                    "Model artifacts not cached; skipping integration test.")
  testthat::skip_if_offline()

  result <- TaxonBodyMassML::predict_mass("Nucella ostrina",
                                          confidence_interval = TRUE)
  expect_true(all(c("lower_bound", "upper_bound", "confidence") %in% names(result)))
  expect_true(is.na(result$lower_bound))
  expect_true(is.na(result$upper_bound))
  expect_true(is.na(result$confidence))
})

test_that("predict_mass() with include_source returns tbmML_genus for model-inferred with known genus", {
  testthat::skip_if(!TaxonBodyMassML:::.artifacts_cached(),
                    "Model artifacts not cached; skipping integration test.")
  testthat::skip_if_offline()

  # Nucella lima is not in training data but Nucella genus is known
  result <- TaxonBodyMassML::predict_mass("Nucella lima", include_source = TRUE)
  expect_true("source" %in% names(result))
  expect_equal(result$source, "tbmML_genus")
})

test_that("predict_mass() with include_source returns NA source for unresolvable taxon", {
  testthat::skip_if(!TaxonBodyMassML:::.artifacts_cached(),
                    "Model artifacts not cached; skipping integration test.")
  testthat::skip_if_offline()

  suppressWarnings(
    result <- TaxonBodyMassML::predict_mass("Xyzzy_definitely_not_a_species_12345",
                                            include_source = TRUE)
  )
  expect_true("source" %in% names(result))
  expect_true(is.na(result$source))
})

test_that("predict_mass() preserves order with mixed dict/model/unresolved rows", {
  testthat::skip_if(!TaxonBodyMassML:::.artifacts_cached(),
                    "Model artifacts not cached; skipping integration test.")
  testthat::skip_if_offline()

  sp <- c("Nucella ostrina",                    # dict hit
          "Canis lupus",                         # model-inferred
          "Xyzzy_definitely_not_a_species_12345") # unresolvable
  suppressWarnings(result <- TaxonBodyMassML::predict_mass(sp, include_source = TRUE))
  expect_equal(nrow(result), 3L)
  expect_equal(result$taxon[1L], "Nucella ostrina")
  expect_equal(result$mass_g[1L], 0.7)
  expect_equal(result$source[1L], "Novak_unpubl")
  expect_false(is.na(result$mass_g[2L]))
  expect_true(startsWith(result$source[2L], "tbmML_"))
  expect_true(is.na(result$mass_g[3L]))
  expect_true(is.na(result$source[3L]))
})

# ---------------------------------------------------------------------------
# lookup parameter
# ---------------------------------------------------------------------------

test_that("predict_mass() with lookup = FALSE routes dict species through model", {
  testthat::skip_if(!TaxonBodyMassML:::.artifacts_cached(),
                    "Model artifacts not cached; skipping integration test.")
  testthat::skip_if_offline()

  # Nucella ostrina is in training data (mass_g = 0.7); with lookup = FALSE
  # it should be passed through the model and return a different value.
  result <- TaxonBodyMassML::predict_mass("Nucella ostrina", lookup = FALSE)
  expect_false(isTRUE(all.equal(result$mass_g, 0.7)))
})

test_that("predict_mass() with lookup = FALSE, include_source returns tbmML_ prefix", {
  testthat::skip_if(!TaxonBodyMassML:::.artifacts_cached(),
                    "Model artifacts not cached; skipping integration test.")
  testthat::skip_if_offline()

  result <- TaxonBodyMassML::predict_mass("Nucella ostrina",
                                          lookup = FALSE,
                                          include_source = TRUE)
  expect_true(startsWith(result$source, "tbmML_"))
})

# ---------------------------------------------------------------------------
# interval_method
# ---------------------------------------------------------------------------

test_that("predict_mass() stratified interval returns finite bounds for model-inferred taxon", {
  testthat::skip_if(!TaxonBodyMassML:::.artifacts_cached(),
                    "Model artifacts not cached; skipping integration test.")
  testthat::skip_if_offline()

  result <- TaxonBodyMassML::predict_mass(
    "Nucella lima",
    confidence_interval = TRUE,
    interval_method = "stratified"
  )
  expect_true(all(c("lower_bound", "upper_bound", "confidence") %in% names(result)))
  expect_true(is.finite(result$lower_bound))
  expect_true(is.finite(result$upper_bound))
  expect_lt(result$lower_bound, result$mass_g)
  expect_gt(result$upper_bound, result$mass_g)
})

test_that("predict_mass() pooled interval returns finite bounds for model-inferred taxon", {
  testthat::skip_if(!TaxonBodyMassML:::.artifacts_cached(),
                    "Model artifacts not cached; skipping integration test.")
  testthat::skip_if_offline()

  result <- TaxonBodyMassML::predict_mass(
    "Nucella lima",
    confidence_interval = TRUE,
    interval_method = "pooled"
  )
  expect_true(is.finite(result$lower_bound))
  expect_true(is.finite(result$upper_bound))
  expect_lt(result$lower_bound, result$mass_g)
  expect_gt(result$upper_bound, result$mass_g)
})

test_that("predict_mass() stratified and pooled intervals differ for model-inferred taxon", {
  testthat::skip_if(!TaxonBodyMassML:::.artifacts_cached(),
                    "Model artifacts not cached; skipping integration test.")
  testthat::skip_if_offline()

  r_strat <- TaxonBodyMassML::predict_mass(
    "Nucella lima", confidence_interval = TRUE, interval_method = "stratified"
  )
  r_pool  <- TaxonBodyMassML::predict_mass(
    "Nucella lima", confidence_interval = TRUE, interval_method = "pooled"
  )
  width_strat <- r_strat$upper_bound - r_strat$lower_bound
  width_pool  <- r_pool$upper_bound  - r_pool$lower_bound
  expect_false(isTRUE(all.equal(width_strat, width_pool)))
})

test_that("predict_mass() dict hit has NA bounds regardless of interval_method", {
  testthat::skip_if(!TaxonBodyMassML:::.artifacts_cached(),
                    "Model artifacts not cached; skipping integration test.")
  testthat::skip_if_offline()

  r_strat <- TaxonBodyMassML::predict_mass(
    "Nucella ostrina", confidence_interval = TRUE, interval_method = "stratified"
  )
  expect_true(is.na(r_strat$lower_bound))
  expect_true(is.na(r_strat$upper_bound))
})
